### PR TITLE
[SPARK-47081][CONNECT] Support Query Execution Progress

### PR DESCRIPTION
**Pull request by grundprinzip**
### What changes were proposed in this pull request?
This patch adss a new mechanism to push query execution progress for batch queries. We add a new response message type and periodically push query progress to the client. The client can consume this data to for example display a progress bar.

This patch adds support for displaying a progress bar in the PySpark shell when started with Spark Connect.

The proto message is defined as follows:

```
// This message is used to communicate progress about the query progress during the execution.
  // This message is used to communicate progress about the query progress during the execution.
  message ExecutionProgress {
    // Captures the progress of each individual stage.
    repeated StageInfo stages = 1;

    // Captures the currently in progress tasks.
    int64 num_inflight_tasks = 2;

    message StageInfo {
      int64 stage_id = 1;
      int64 num_tasks = 2;
      int64 num_completed_tasks = 3;
      int64 input_bytes_read = 4;
      bool done = 5;
    }
  }
```

Clients can simply ignore the messages or consume them. On top of that this adds additional capabilities to register a callback for progress tracking to the SparkSession.

```
handler = lambda **kwargs: print(kwargs)
spark.register_progress_handler(handler)
spark.range(100).collect()
spark.remove_progress_handler(handler)
```

#### Example 1
progress_medium_query_multi_stage mp4

#### Example 2
progress_bar mp4

### Why are the changes needed?
Usability and Experience

### Does this PR introduce _any_ user-facing change?
When the user opens the PySpark shell with Spark Connect mode, it will use the progress bar by default.

### How was this patch tested?
Added new tests.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: python/pyspark/shell.py [hunk 1]
```diff
@@ -45,11 +45,29 @@
         if parent_dir in sys.path:
             sys.path.remove(parent_dir)
 
-
 if is_remote():
     try:
         # Creates pyspark.sql.connect.SparkSession.
         spark = SparkSession.builder.getOrCreate()
```
Comment (HyukjinKwon) [inline python/pyspark/shell.py:51]: Could we get this from `spark.conf`? I think we should avoid relying on environment variable in general. Should probably also disable `spark.ui.showConsoleProgress` here.

File: connector/connect/common/src/main/protobuf/spark/connect/base.proto [hunk 2]
```diff
@@ -420,6 +423,15 @@ message ExecutePlanResponse {
     // the execution is complete. If the server sends onComplete without sending a ResultComplete,
     // it means that there is more, and the client should use ReattachExecute RPC to continue.
   }
+
+  // This message is used to communicate progress about the query progress during the execution.
+  message ExecutionProgress {
+    int64 num_tasks = 1;
+    int64 num_completed_tasks = 2;
+    int64 num_stages = 3;
+    int64 num_completed_stages = 4;
```
Comment (HyukjinKwon) [inline connector/connect/common/src/main/protobuf/spark/connect/base.proto:432]: We're carrying the state and task information but I can't see in the PR description example. How are they displayed?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+import org.apache.spark.scheduler.{SparkListener, SparkListenerJobEnd, SparkListenerJobStart, SparkListenerStageCompleted, SparkListenerTaskEnd}
+
+/**
+ * A listener that tracks the execution of jobs and stages for a given set of tags. This is used
+ * to track the progress of a job that is being executed through the connect API.
+ *
+ * The listener is instantiated once for the SparkConnectService and then used to track all the
+ * current query executions.
+ */
+private[connect] class ConnectProgressExecutionListener extends SparkListener with Logging {
```
Comment (HyukjinKwon) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala:30]: Could we reuse `AppStatusListener` ? That is used by non Spark Connect code path (see `ConsoleProgressBar`)

Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala:30]: The `ConsoleProgressBar` "randomly" shows the progress of up to three stages. The stages can be completely independent of the actual query execution of this Spark Session. This is the reason why I created a custom listener that only tracks Spark Connect queries for the whole SparkConnect service and collapses them by operation tag.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    def jobCount(): Int = {
+      jobs.size
+    }
+
+    def stageCount(): Int = {
+      stages.size
+    }
+  }
+
+  val trackedTags = collection.mutable.Map[String, ExecutionTracker]()
```
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala:81]: This needs to be thread safe. This is both accessed from the connect execution thread, and the listener bus thread.

Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala]: Done

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  override def onStageCompleted(stageCompleted: SparkListenerStageCompleted): Unit = {
+    trackedTags.foreach({ case (tag, tracker) =>
+      if (tracker.stages.contains(stageCompleted.stageInfo.stageId)) {
+        tracker.completedStages += 1
+        tracker.dirty = true
+      }
+    })
+  }
+
+  override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
```
[the remark this answers is not in the document]
Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala:124]: you mean for reporting or in general?

**Comment by grundprinzip** [comment]
Any chance to get some more feedback here? @HyukjinKwon or @hvanhovell ?

**Comment by dtenedor** [comment]
cc @ueshin @cloud-fan we need help 🙏

File: connector/connect/common/src/main/protobuf/spark/connect/base.proto [hunk 6]
```diff
@@ -435,6 +438,16 @@ message ExecutePlanResponse {
...
+    int64 input_bytes_read = 5;
+    int64 num_inflight_tasks = 6;
```
Comment (cloud-fan) [inline connector/connect/common/src/main/protobuf/spark/connect/base.proto:449]: when do we need this info?

File: connector/connect/common/src/main/protobuf/spark/connect/base.proto [hunk 7]
```diff
@@ -435,6 +438,16 @@ message ExecutePlanResponse {
     // the execution is complete. If the server sends onComplete without sending a ResultComplete,
     // it means that there is more, and the client should use ReattachExecute RPC to continue.
   }
+
+  // This message is used to communicate progress about the query progress during the execution.
+  message ExecutionProgress {
+    int64 num_tasks = 1;
+    int64 num_completed_tasks = 2;
```
[the remark this answers is not in the document]
Comment (grundprinzip) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto:445]: Across all stages. It can always be extended later.

[the remark this answers is not in the document]
Comment (cloud-fan) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto:445]: I'm wondering how can this be accurate. With AQE we never know what is the number of partitions for the next stage, as re-optimization can happen.

[the remark this answers is not in the document]
Comment (grundprinzip) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto:445]: The goal of the progress metrics is not to be accurate into the future but only represent the snapshot of the current state. This means that the number of tasks can be updated when new stages are added or AQE kicks in.

The point is that the number of remaining tasks will converge over time and become stable.

File: connector/connect/common/src/main/protobuf/spark/connect/base.proto (hunk 6 above)
Comment (grundprinzip) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto:449]: This is used in the progress bar implementation.

File: connector/connect/common/src/main/protobuf/spark/connect/base.proto (hunk 7 above)
[the remark this answers is not in the document]
Comment (grundprinzip) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto:445]: progres

[the remark this answers is not in the document]
Comment (cloud-fan) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto:445]: This looks a bit different from what we have in classic Spark now. Classic Spark refreshes the progress bar after one stage is finished. So people may see something like below at the beginning
```
[Stage 1:=================================>                 (6619 + 12) / 10000]
```
then after stage 1 is completed, the progress bar is refreshed
```
[Stage 2:=====================>                             (3214 + 12) / 8000]
```

We don't have to follow the exact display format, but this "stage by stage" approach is more nature fit of the actual query execution in Spark.

[the remark this answers is not in the document]
Comment (grundprinzip) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto:445]: I think you're overindexing on accuracy. The progress bar is an ephemeral piece that simply gives the user feedback about the current operations happening. This is true for the current progress bar as well it shows a current stage and will delete it when it gets canceled.

To the user, it does not matter if the percentage is accurate or not as long as it converges toward 100% at some point in time. Simply seeing that any progress is happening and the concurrency of it is a benefit to the user.

[the remark this answers is not in the document]
Comment (cloud-fan) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto:445]: I agree we don't have to be very accurate, but is there a reason that we need to have a different progress bar style for Spark Connect? Is it because the classic Spark "stage by stage" progress bar is bad, or it's hard to implement?

[the remark this answers is not in the document]
Comment (aarondav) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto:445]: (Just my 2c: I think having any progress bar is much better than none. The standard Spark progress bar has some ups and some downs, definitely having new progress bars appear isn't the most intuitive either. I think it's probably net better than one progress bar that gets longer, but I would much prefer having some progress bar now that we can extend later, perhaps as we get a better sense of how to incorporate AQE and future stages into the UX.)

[the remark this answers is not in the document]
Comment (cloud-fan) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto:445]: After a second thought, it's better to hide Spark internals (stages) to end users, and eventually we should only have one progress bar for the query. So the current PR is a good starting point.

However, this server-client protocol needs to be stable and we don't want to change the client frequently to improve the progress reporting. Can we define a minimum set of information we need to send to the client side to display the progress bar? I feel it's better to calculate the percentage at the server side.

[the remark this answers is not in the document]
Comment (grundprinzip) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto]: So I refactored the code to avoid closing any doors. I did not change the way the progress bar is displayed. However, I extended the progress message to capture the stage-wise information so other clients can decide independently how to present the information to the end user.

[the remark this answers is not in the document]
Comment (aarondav) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto]: +1 @cloud-fan what do you think about that? Capture stage-level info in the proto, but keep the display simple for now?

File: python/pyspark/sql/connect/session.py [hunk 8]
```diff
@@ -326,6 +327,28 @@ def readStream(self) -> "DataStreamReader":
 
     readStream.__doc__ = PySparkSession.readStream.__doc__
 
+    def registerProgressHandler(self, handler: "ProgressHandler") -> None:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/session.py:330]: nit but API should be put in `python/docs/source/reference/pyspark.sql/spark_session.rst`

File: python/pyspark/sql/connect/session.py [hunk 9]
```diff
@@ -326,6 +327,28 @@ def readStream(self) -> "DataStreamReader":
...
+        """
+        Register a progress handler to be called when a progress update is received from the server.
+        .. versionadded:: 4.0
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/session.py:333]: API should follow https://numpydoc.readthedocs.io/en/latest/format.html with describing return type, and example if possible.

File: python/pyspark/sql/connect/session.py [hunk 10]
```diff
@@ -1007,6 +1030,7 @@ def profile(self) -> Profile:
 
 def _test() -> None:
     import sys
+    import os  # noqa: F401
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/session.py:1033]: nit but should probably remove this

Comment (grundprinzip) [reply python/pyspark/sql/connect/session.py:1033]: I can't remove this because the doc tests fail otherwise.

Comment (HyukjinKwon) [reply python/pyspark/sql/connect/session.py:1033]: which doc test fails? I see no new tests added in this file that uses `os`.

File: python/pyspark/sql/connect/shell/progress.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+        char: str = "*",
+        min_width: int = 80,
+        output: typing.IO = sys.stdout,
+        enabled: bool = False,
+        handlers: Iterable[ProgressHandler] = [],
+    ) -> None:
+        """
+        Constructs a new Progress bar. The progress bar is typically used in
+        the blocking query execution path to process the execution progress
+        methods from the server.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/shell/progress.py:95]: should have a newline here otherwise the html output is malformed.

File: python/pyspark/sql/connect/shell/progress.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+                done=done,
+            )
+
+    def update_ticks(self, stages: Iterable[StageInfo], inflight_tasks: int) -> None:
+        """This method is called from the execution to update the progress bar with a new total
+        tick counter and the current position. This is necessary in case new stages get added with
+        new tasks and so the total task number will be updated as well.
+
+        Parameters
+        ==========
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/shell/progress.py:133]: should use `-`.

File: python/pyspark/sql/connect/session.py (hunk 10 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/session.py:1033]: A i removed the example because it failed in weird modes.

I had

```
h = lambda **kwargs: print(kwargs)
spark.registerProgressHandler(h)
```

Then it failed with print not found. I changed it to os.write but then os was not found. This got fixed by importing os. 

But the example becomes too complicated.

File: python/pyspark/sql/connect/client/core.py [hunk 13]
```diff
@@ -1335,6 +1370,16 @@ def handle_response(
...
                         for b in self._stub.ExecutePlan(req, metadata=self._builder.metadata()):
                             yield from handle_response(b)
+        except KeyboardInterrupt:
+            logger.debug(f"Interrupt request received for operation={req.operation_id}")
+            try:
+                self.interrupt_operation(req.operation_id)
+            except Exception as e:
+                # Swallow all errors if aborted.
+                logger.debug(f"Caught an error during interrupt handling, silenced: {e}")
+                pass
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/client/core.py:1380]: nit remove

File: python/pyspark/sql/connect/client/core.py [hunk 14]
```diff
@@ -694,6 +695,33 @@ class ClientThreadLocals(threading.local):
 
         self._profiler_collector = ConnectProfilerCollector()
 
+        self._progress_handlers: List[ProgressHandler] = []
+
+    def register_progress_handler(self, handler: ProgressHandler) -> None:
+        """
+        Register a progress handler to be called when a progress message is received.
+        Parameters
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/client/core.py:703]: newline before `Parameters`.

File: python/pyspark/shell.py [hunk 15]
```diff
@@ -45,11 +45,30 @@
         if parent_dir in sys.path:
             sys.path.remove(parent_dir)
 
-
```
Comment (HyukjinKwon) [inline python/pyspark/shell.py:48]: I didn't want to point this out but since it needs some changes .. I would revert this change back

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+ * The listener is instantiated once for the SparkConnectService and then used to track all the
+ * current query executions.
+ */
+private[connect] class ConnectProgressExecutionListener extends SparkListener with Logging {
+
+  /**
+   * A tracker for a given tag. This is used to track the progress of an operation is being
+   * executed through the connect API.
+   */
+  class ExecutionTracker(var tag: String) {
```
Comment (HyukjinKwon) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala:39]: maybe I missed some context but why is this `var`?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 17]
```diff
@@ -743,8 +743,11 @@ private[spark] class SparkSubmit extends Logging {
       }
     }
 
-    // In case of shells, spark.ui.showConsoleProgress can be true by default or by user.
-    if (isShell(args.primaryResource) && !sparkConf.contains(UI_SHOW_CONSOLE_PROGRESS)) {
+    // In case of shells, spark.ui.showConsoleProgress can be true by default or by user. Except,
+    // when Spark Connect is in local mode, because Spark Connect support its own progress
+    // reporting.
+    if (isShell(args.primaryResource) && !sparkConf.contains(UI_SHOW_CONSOLE_PROGRESS) &&
+      !sparkConf.contains("spark.local.connect")) {
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:750]:
```suggestion
        !sparkConf.contains("spark.local.connect")) {
```

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala [hunk 18]
```diff
@@ -265,4 +265,11 @@ object Connect {
       .version("4.0.0")
       .bytesConf(ByteUnit.BYTE)
       .createWithDefault(1024)
+
+  val CONNECT_PROGRESS_REPORT_INTERVAL =
```
Comment (zhengruifeng) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala:269]: Does it make sense to make it a runtime config?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala [hunk 19]
```diff
@@ -265,4 +265,11 @@ object Connect {
...
+    buildStaticConf("spark.connect.progress.reportInterval")
+      .doc("The interval at which the progress of a query is reported to the client.")
+      .version("4.0.0")
+      .timeConf(TimeUnit.MILLISECONDS)
+      .createWithDefaultString("2s")
```
Comment (zhengruifeng) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala:274]: What about making this config accept '-1' value? and disable this feature if the value is negative

File: python/pyspark/sql/connect/session.py (hunk 10 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/session.py:1033]: Updating with examples.

File: python/pyspark/sql/connect/session.py (hunk 8 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/session.py:330]: Done

File: python/pyspark/sql/connect/shell/progress.py (hunk 12 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/shell/progress.py:133]: Done

File: python/pyspark/sql/tests/connect/shell/test_progress.py [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+from io import StringIO
+import unittest
+from typing import Iterable
+
+from pyspark.testing.connectutils import (
+    should_test_connect,
+    connect_requirement_message,
+)
+from pyspark.testing.utils import PySparkErrorTestUtils
+from pyspark.sql.connect.shell.progress import Progress, StageInfo
```
[the remark this answers is not in the document]
Comment (grundprinzip) [reply python/pyspark/sql/tests/connect/shell/test_progress.py:27]: Done

File: python/pyspark/sql/connect/client/core.py (hunk 13 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/client/core.py:1380]: since this is only debug logging, might be worthwhile to keep to make it easier to debug special cases where the keyboard interrupt is fired and the query cancellation fails?

File: python/pyspark/sql/connect/client/core.py (hunk 14 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/client/core.py:703]: Done

File: python/pyspark/shell.py (hunk 15 above)
Comment (grundprinzip) [reply python/pyspark/shell.py:48]: Done

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 17 above)
Comment (grundprinzip) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:750]: So I'm not sure this is the proper indentation, my intellij with the scalafmt profile says it is correct like this 🤷 Yay for auto format

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala (hunk 19 above)
Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala:274]: Good idea!

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala (hunk 18 above)
Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/config/Connect.scala:269]: changed it.

**Comment by grundprinzip** [comment]
@zhengruifeng @HyukjinKwon addressed your comments, please take another look.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+          .setNumCompletedTasks(completedTasks)
+          .setInputBytesRead(inputBytesRead)
+          .setDone(completed)
+          .build()
+      }
+    }
+
+    // The set of jobs that are being tracked by this tracker. We always only add to this list
+    // but never remove. This is to avoid concurrency issues.
+    private[ConnectProgressExecutionListener] var jobs: Set[Int] = Set()
```
Comment (cloud-fan) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala:69]: nit: `private[this]` is the narrowest scope we can use here.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+        }
+        tracker.dirty.set(true)
+      }
+    })
+  }
+
+  override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
+    trackedTags.foreach({ case (_, tracker) =>
+      if (tracker.jobs.contains(jobEnd.jobId)) {
+        tracker.dirty.set(true)
```
Comment (cloud-fan) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala:170]: why do we set the dirty flag when nothing is updated?

File: connector/connect/common/src/main/protobuf/spark/connect/base.proto (hunk 7 above)
[the remark this answers is not in the document]
Comment (cloud-fan) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto]: yea this is more flexible. The proto message contains all the information and clients can do whatever they want.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala (hunk 22 above)
Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala:170]: This is mostly to make sure that all progress is reported and an update is sent to the client. If you're tracking time between progress messages, every message itself is progress.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala (hunk 21 above)
Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala:69]: Fixing it.

Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala:69]: Actually, this does not work here because `this` is not the right scope and gives compile errors.

**Comment by grundprinzip** [comment]
@HyukjinKwon @zhengruifeng @cloud-fan I addressed the commments, is there additional feedback?

**Review by HyukjinKwon — COMMENTED** [review]
The logic seems fine. For the output shape and information, would be great if someone like @cloud-fan @hvanhovell reviews it.

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: python/pyspark/sql/session.py [hunk 23]
```diff
@@ -2002,6 +2003,61 @@ def addArtifacts(
...
+
+        Examples
+        --------
+
+        >>> def progress_handler(stages, inflight_tasks, done):
+        ...     print(f"{len(stages)} Stages known, Done: {done}")
+        >>> spark.registerProgressHandler(progress_handler)
+        >>> res = spark.range(10).repartition(1).collect()
+        3 Stages known, Done: False
+        3 Stages known, Done: True
```
Comment (HyukjinKwon) [inline python/pyspark/sql/session.py:2026]: This test is flaky:

```
File "/__w/spark/spark/python/pyspark/sql/connect/session.py", line 346, in pyspark.sql.connect.session.SparkSession.registerProgressHandler
Failed example:
    res = spark.range(10).repartition(1).collect()
Expected:
    3 Stages known, Done: False
    3 Stages known, Done: True
Got:
    0 Stages known, Done: True
```

https://github.com/apache/spark/actions/runs/8564043093/job/23470007059.

Let me skip it for now.

Comment (grundprinzip) [reply python/pyspark/sql/session.py:2026]: I'll unflake it. Thanks!

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala [hunk 24]
```diff
@@ -201,9 +237,18 @@ private[connect] class ExecuteGrpcResponseSender[T <: Message](
...
-            val timeout = Math.max(1, deadlineTimeMillis - System.currentTimeMillis())
+            // Wake up more frequently to send the progress updates.
+            val progressTimeout =
+              executeHolder.sessionHolder.session.conf.get(CONNECT_PROGRESS_REPORT_INTERVAL)
+            // If the progress feature is disabled, wait for the deadline.
+            val timeout = if (progressTimeout > 0) {
+              progressTimeout
+            } else {
+              Math.max(1, deadlineTimeMillis - System.currentTimeMillis())
+            }
```
Comment (juliuszsompolski) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala:248]: nit:
```
            var timeout = Math.max(1, deadlineTimeMillis - System.currentTimeMillis())
            // Wake up more frequently to send the progress updates.
            val progressTimeout =
              executeHolder.sessionHolder.session.conf.get(CONNECT_PROGRESS_REPORT_INTERVAL)
            if (progressTimeout > 0) {
              Math.min(progressTimeout, timeout)
            }
```
otherwise, progressTimeout may make us wait beyond the deadline.

Comment (HyukjinKwon) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteGrpcResponseSender.scala:248]: https://github.com/apache/spark/pull/49474

**Comment by virrrat** [comment]
Is there a plan to port-back this feature to Spark 3.5?

**Comment by grundprinzip** [comment]
@virrrat this is not planned and given the Spark 4 release soon, not practical.

File: python/pyspark/sql/session.py [hunk 25]
```diff
@@ -2002,6 +2003,61 @@ def addArtifacts(
 
     addArtifact = addArtifacts
 
+    def registerProgressHandler(self, handler: "ProgressHandler") -> None:
+        """
+        Register a progress handler to be called when a progress update is received from the server.
+
+        .. versionadded:: 4.0
```
Comment (nchammas) [inline python/pyspark/sql/session.py:2010]: Potentially silly question, but: When you look at [the docs for this][1], it's not obvious that Spark Connect supports this method. Should this be explicitly noted in the docstring somehow? Or are users supposed to assume that everything supports Spark Connect unless explicitly noted otherwise?

[1]: https://spark.apache.org/docs/4.0.0/api/python/reference/pyspark.sql/api/pyspark.sql.SparkSession.registerProgressHandler.html?highlight=progress#pyspark.sql.SparkSession.registerProgressHandler

Comment (nchammas) [reply python/pyspark/sql/session.py:2010]: Another related question: Should there be narrative documentation of `ProgressHandler` on the [monitoring page][1], or are we happy with it just being tucked away in the API docs?

[1]: https://spark.apache.org/docs/4.0.0/monitoring.html

File: python/pyspark/sql/connect/shell/progress.py [hunk 26]
```diff
@@ -0,0 +1,187 @@
...
+from pyspark.sql.connect.shell import progress_bar_enabled
+
+
+@dataclass
+class StageInfo:
+    stage_id: int
+    num_tasks: int
+    num_completed_tasks: int
+    num_bytes_read: int
+    done: bool
```
Comment (nchammas) [inline python/pyspark/sql/connect/shell/progress.py:45]: I'm just learning about this new feature and am potentially interested in expanding on the documentation for it, as it's very useful for people building applications on top of Spark.

One thing I noticed is that a job can be marked as "done" even though the number of completed tasks is less than the number of tasks for one or more stages. I assume this is because the stage was skipped or something else, but this information is not captured in this class, so the resulting progress communicated to the user ends up being a bit misleading and/or noisy.

Is it possible to enhance this somehow with that information (in which case I'm happy to file a ticket), or have I misunderstood this data?

Here's an example from some testing I did. This is the last update I got from my progress handler:

```python
{'stages': [StageInfo(stage_id=37,
                      num_tasks=1,
                      num_completed_tasks=1,
                      num_bytes_read=0,
                      done=True),
            StageInfo(stage_id=29,
                      num_tasks=1,
                      num_completed_tasks=1,
                      num_bytes_read=0,
                      done=True),
            StageInfo(stage_id=33,
                      num_tasks=183,
                      num_completed_tasks=183,
                      num_bytes_read=0,
                      done=True),
            StageInfo(stage_id=35,
                      num_tasks=120,
                      num_completed_tasks=0,
                      num_bytes_read=0,
                      done=False),
            StageInfo(stage_id=31,
                      num_tasks=1,
                      num_completed_tasks=0,
                      num_bytes_read=0,
                      done=False),
            StageInfo(stage_id=32,
                      num_tasks=120,
                      num_completed_tasks=0,
                      num_bytes_read=0,
                      done=False),
            StageInfo(stage_id=34,
                      num_tasks=1,
                      num_completed_tasks=0,
                      num_bytes_read=0,
                      done=False),
            StageInfo(stage_id=36,
                      num_tasks=183,
                      num_completed_tasks=0,
                      num_bytes_read=0,
                      done=False),
            StageInfo(stage_id=30,
                      num_tasks=120,
                      num_completed_tasks=120,
                      num_bytes_read=0,
                      done=True)],
 'inflight_tasks': 0,
 'operation_id': '1a9fbf1d-4a38-4c6b-b730-6c8b49179694',
 'done': True}
```

Note that the overall status is "done", even though many stages are not themselves done.

File: python/pyspark/sql/session.py (hunk 25 above)
Comment (grundprinzip) [reply python/pyspark/sql/session.py:2010]: Creating a PR with documentation updates would be very much appreciated!

File: python/pyspark/sql/connect/shell/progress.py (hunk 26 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/shell/progress.py:45]: The progress reporting of the query is meant to provide current information about the query with the progress message continuing until the query is observed as done from the client side. 

Due to the way that the Spark event listeners work, it is not guaranteed that all events have been processed until the query is marked as done by the Spark Connect query execution.

This means there are two potential ways the data can be "off":

1. Skipped / Canceled stages
2. Events not yet processed.

The goal for the progress report is not to be 100% accurate but indicate to the user what kind of progress the operation is making, for that reason the completed and current task might not be a perfect measure but provide a reasonable approximation that converges to a reasonable progress report.

Comment (virrrat) [reply python/pyspark/sql/connect/shell/progress.py:45]: Would it makes sense to add the `stage_status` field to `StageInfo` to better handle skipped / cancelled stages scenario? Users can consume data based on status

Comment (nchammas) [reply python/pyspark/sql/connect/shell/progress.py:45]: The Spark UI does document skipped stages, but I'm not sure if that information is available before the job is done. I think that's what @grundprinzip is saying.
